pkg/storage/parse.go

```diff
@@ -0,0 +1,89 @@
+package storage
+
+import (
+	"net/url"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/backup"
+)
+
+// BackendOptions further configures the storage backend not expressed by the
+// storage URL.
+type BackendOptions struct {
+	S3 S3BackendOptions `json:"s3" toml:"s3"`
+}
+
+func (options *BackendOptions) Validate(storageScheme string) error {
+	optionsNotApplicableToStorage := func(option string) error {
+		return errors.Errorf("options '%s.*' are not applicable to %s storage", option, storageScheme)
+	}
+
+	if !options.S3.isEmpty() && storageScheme != "s3" {
+		return optionsNotApplicableToStorage("s3")
+	}
+	if err := options.S3.validate(); err != nil {
+		return err
+	}
+
+	return nil
```

Comment:
This is not straightforward, if scheme=="s3", we only need to care about if options.S3 is empty, because even if both options.S3 and options.GCS are not empty, we can still work as expected. 
Can we make it like below:
```
switch storageScheme{
case s3:
  if options.S3.isEmpty(){
    return error
  }
  validate()
```
If the above is accepted, we may just migrate the logic to the blocks from L48.

Reply:
The point of the check is that if you have `--storage local://...` and provide an --s3 option it is invalid. So this doesn't matter?

Reply:
It does matter, I mean we should check the options by scheme instead of checking the scheme by the options, for your example, if --storage local://... and provide both --s3 options and the local options (if some), BR can work with local:// and the local options, right? No need to return an error just because users also provide s3 options.
It should be more clear with this case:
--storage s3://...
Users provide both s3 options and gcs options
I think no need to return an error because BR can just use the s3 options to work.

Reply:
With this argument we should allow `--storage local://… --s3.endpoint … --local.stuff …`, but since local has no options in reality we should just allow `--storage local://… --s3.endpoint …` 

The missing s3 options could cause validation failure if we require either --s3.region or --s3.endpoint must be present with an s3 storage. This is a different issue from using X options on Y storage.

Reply:
See the first [comment](https://github.com/pingcap/br/pull/88#discussion_r355420437), we only need to check s3 options when the scheme is s3.